rest/diagnostic_api.go

```diff
@@ -0,0 +1,222 @@
+//  Copyright 2013-Present Couchbase, Inc.
+//
+//  Use of this software is governed by the Business Source License included
+//  in the file licenses/BSL-Couchbase.txt.  As of the Change Date specified
+//  in that file, in accordance with the Business Source License, use of this
+//  software will be governed by the Apache License, Version 2.0, included in
+//  the file licenses/APL2.txt.
+
+package rest
+
+import (
+	"fmt"
+
+	"github.com/couchbase/sync_gateway/auth"
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/couchbase/sync_gateway/channels"
+	"github.com/gorilla/mux"
+)
+
+// each field in this response is a map of collection names to a map of channel names to grant history info
+type getAllChannelsResponse struct {
+	AdminGrants       map[string]map[string]auth.GrantHistory            `json:"admin_grants,omitempty"`
+	DynamicGrants     map[string]map[string]auth.GrantHistory            `json:"dynamic_grants,omitempty"`
+	JWTGrants         map[string]map[string]auth.GrantHistory            `json:"jwt_grants,omitempty"`
+	AdminRoleGrants   map[string]map[string]map[string]auth.GrantHistory `json:"admin_role_grants,omitempty"`
+	DynamicRoleGrants map[string]map[string]map[string]auth.GrantHistory `json:"dynamic_role_grants,omitempty"`
+}
+
+func newGetAllChannelsResponse() getAllChannelsResponse {
+	var resp getAllChannelsResponse
+	resp.AdminGrants = make(map[string]map[string]auth.GrantHistory)
+	resp.DynamicGrants = make(map[string]map[string]auth.GrantHistory)
+	resp.DynamicRoleGrants = make(map[string]map[string]map[string]auth.GrantHistory)
+	resp.AdminRoleGrants = make(map[string]map[string]map[string]auth.GrantHistory)
+	return resp
+}
+
+func (h *handler) handleGetAllChannels() error {
+	h.assertAdminOnly()
+	username := mux.Vars(h.rq)["name"]
+	user, err := h.db.Authenticator(h.ctx()).GetUser(internalUserName(username))
+	if err != nil {
+		return fmt.Errorf("Could not get user %s: %w", username, err)
+	}
+	if user == nil {
+		return fmt.Errorf("Could not get user %s: %w", username, err)
+	}
+
+	authenticator := h.db.Authenticator(h.ctx())
```

Comment:
```suggestion
	authenticator := h.db.Authenticator(h.ctx())
	username := mux.Vars(h.rq)["name"]
	user, err := authenticator.GetUser(internalUserName(username))
	if err != nil {
		return fmt.Errorf("Error getting user %s: %w", username, err)
	}
	if user == nil {
		return base.HTTPErrorf(http.StatusNotFound, "User not found %s", username)
	}
```

Since we will use authenticator repeatedly, let's define it up front. The error for the first case contains an error that is probably typed, but if `user == nil` there is no `err` so we'll make our own message. Making the text distinct from above means that we are likely to be able to find it if a user hits the problem.